AK/SinglyLinkedList.h

```diff
@@ -128,6 +128,60 @@ class SinglyLinkedList {
         m_tail->next = node;
         m_tail = node;
     }
+    
+    void sorted_insert_slow(const T& value)
+    {
+        auto* new_node = new Node(move(value));
+        new_node->value = value;
+
+        if (!m_head) {
+            m_head = new_node;
+            m_tail = new_node;
+            return;
+        }
+
+	if (!m_head->next && new_node->value < m_head->value) {
+           new_node->next = m_head;
+	   m_head = new_node;
+	   return;
+	}
+
+        Node* curr = m_head;
+	while (curr->next && curr->next->value < new_node->value) 
+	{
+	    curr = curr->next;
+        }
+
+        new_node->next = curr->next;
+	curr->next = new_node;
+	if(m_tail == curr)
+            m_tail = new_node;
+
+    }   
+    
+    void sorted_insert_slow(T&& value)
```

Comment:
It really seems like this should be a separate data structure. It could be a wrapper around a `SinglyLinkedList` at first, if you like.

That way the new class can make sure to only provide API's that keep the list sorted, while the old class can stay unsorted. We should not force clients of the class to know whether an instance maintains sort order or not.

For example, it could be a new `SortedList` template class in AK, or a kernel-specific `SortedTimerList` for example. (The names are just ideas.)

Reply:
That sounds reasonable. If we'd go the kernel-specific route, wouldn't it be better to have a static `TimerQueue` class? This could do the sorting of a `SinglyLinkedList` and also provide methods for adding and deleting timers.

Reply:
Sure, that sounds nice!